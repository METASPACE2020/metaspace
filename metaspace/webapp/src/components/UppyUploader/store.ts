// based on https://uppy.io/docs/stores/#Implementing-Stores

import { ref } from 'vue'
import { State } from '@uppy/core'

type Listener = (prevState: State, nextState: State, patch: object) => void

export default () => {
  const state = ref<State | any>({})
  const listeners = new Set<Listener>()

  return {
    getState: () => state.value,
    setState: (patch: object) => {
      const prevState = state.value
      const nextState = { ...prevState, ...patch }

      state.value = nextState

      listeners.forEach((listener) => {
        listener(prevState, nextState, patch)
      })
    },
    subscribe: (listener: Listener) => {
      listeners.add(listener)
      return () => listeners.delete(listener)
    },
  }
}
