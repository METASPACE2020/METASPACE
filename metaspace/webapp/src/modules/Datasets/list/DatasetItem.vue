<template>
  <div
    v-if="!deferRender"
    class="dataset-item border border-solid border-gray-200 leading-5"
    :class="disabledClass"
  >
    <el-dialog
      title="Provided metadata"
      :lock-scroll="false"
      :visible.sync="showMetadataDialog"
    >
      <dataset-info
        :metadata="metadata"
        :current-user="currentUser"
      />
    </el-dialog>

    <div
      v-if="isOpticalImageSupported"
      class="opt-image--container"
    >
      <dataset-thumbnail
        :dataset="dataset"
        :editable="canEditOpticalImage"
      />
    </div>

    <div class="ds-info">
      <div class="ds-item-line flex">
        <!-- title is set to make it easier to see overflowing datasets' names by hovering over the name -->
        <span
          :title="formatDatasetName"
          class="font-bold truncate"
        >{{ formatDatasetName }}</span>
        <el-popover
          v-if="!dataset.isPublic"
          class="ml-1"
          trigger="hover"
          placement="top"
          @show="loadVisibility"
        >
          <div v-loading="visibilityText == null">
            {{ visibilityText }}
          </div>
          <i
            slot="reference"
            class="el-icon-lock"
          />
        </el-popover>
      </div>

      <div class="ds-item-line text-gray-700">
        <span
          class="ds-add-filter"
          title="Filter by species"
          @click="addFilter('organism')"
        >
          {{ formatOrganism }}</span>,
        <span
          class="ds-add-filter"
          title="Filter by organism part"
          @click="addFilter('organismPart')"
        >
          {{ formatOrganismPart }}</span>

        <span
          class="ds-add-filter"
          title="Filter by condition"
          @click="addFilter('condition')"
        >
          ({{ formatCondition }})</span>
      </div>
      <div class="ds-item-line">
        <span
          class="ds-add-filter"
          title="Filter by ionisation source"
          @click="addFilter('ionisationSource')"
        >
          {{ dataset.ionisationSource }}</span> +
        <span
          class="ds-add-filter"
          title="Filter by analyzer type"
          @click="addFilter('analyzerType')"
        >
          {{ dataset.analyzer.type }}</span>,
        <span
          class="ds-add-filter"
          title="Filter by polarity"
          @click="addFilter('polarity')"
        >
          {{ dataset.polarity.toLowerCase() }} mode</span>,
        RP {{ formatResolvingPower }}
      </div>

      <div class="ds-item-line">
        Submitted <elapsed-time :date="getDate" /> by
        <span
          class="ds-add-filter"
          title="Filter by submitter"
          @click="addFilter('submitter')"
        >
          {{ formatSubmitter }}</span><!--
          Be careful not to add empty space before the comma
          --><span v-if="dataset.groupApproved && dataset.group">,
          <el-dropdown
            :show-timeout="50"
            placement="bottom"
            :trigger="hideGroupMenu ? 'never' : 'hover'"
            @command="handleDropdownCommand"
          >
            <span
              class="text-base text-primary cursor-pointer"
              @click="addFilter('group')"
            >
              {{ dataset.group.shortName }}
            </span>
            <el-dropdown-menu slot="dropdown">
              <el-dropdown-item command="filter_group">Filter by this group</el-dropdown-item>
              <el-dropdown-item command="view_group">View group</el-dropdown-item>
            </el-dropdown-menu>
          </el-dropdown>
        </span>
      </div>
      <div
        v-if="dataset.status == 'FINISHED' && dataset.fdrCounts"
        class="ds-item-line"
      >
        <span>
          <router-link :to="resultsHref(formatDbName())">{{ formatFdrCounts() | plural('annotation', 'annotations') }}</router-link>
          @ FDR {{ formatFdrLevel() }}% ({{ formatDbName() }})
        </span>
      </div>
    </div>

    <div class="ds-actions">
      <span v-if="dataset.status == 'FINISHED'">
        <i class="el-icon-picture" />
        <el-popover
          trigger="hover"
          placement="top"
        >
          <div class="db-link-list">
            Select a database:
            <div
              v-for="database in metaboliteDatabases"
              :key="database"
            >
              <router-link :to="resultsHref(database)">
                {{ database }}
              </router-link>
            </div>
          </div>
          <a slot="reference">Browse annotations</a>
        </el-popover>
        <br>
      </span>

      <span v-if="dataset.status === 'ANNOTATING'">
        <div
          class="striped-progressbar processing"
          title="Processing is under way"
        />
      </span>

      <span v-if="dataset.status === 'QUEUED'">
        <div
          class="striped-progressbar queued"
          title="Waiting in the queue"
        />
      </span>

      <div>
        <i class="el-icon-view" />
        <a
          href="#"
          @click="showMetadata"
        >Show full metadata</a>
      </div>

      <div v-if="canEdit">
        <i class="el-icon-edit" />
        <router-link :to="editHref">
          Edit metadata
        </router-link>
      </div>

      <div
        v-if="dataset.canDownload"
        class="ds-download"
      >
        <i class="el-icon-download" />
        <a
          href="#"
          @click.prevent="() => { showDownloadDialog = true }"
        >Download</a>
      </div>

      <div
        v-if="canDelete"
        class="ds-delete"
      >
        <i class="el-icon-delete" />
        <a
          href="#"
          class="text-danger"
          @click.prevent="openDeleteDialog"
        >Delete dataset</a>
      </div>

      <div
        v-if="canReprocess"
        class="ds-reprocess"
      >
        <i class="el-icon-refresh" />
        <a
          href="#"
          class="text-danger"
          @click.prevent="handleReprocess"
        >Reprocess dataset</a>
      </div>

      <div
        v-else-if="isPublished"
        class="mt-auto text-right text-gray-700 text-sm"
      >
        Published
      </div>
    </div>
    <DownloadDialog
      v-if="showDownloadDialog"
      :dataset-id="dataset.id"
      :dataset-name="dataset.name"
      @close="() => { showDownloadDialog = false }"
    />
  </div>
</template>

<script>
import DatasetInfo from '../../../components/DatasetInfo.vue'
import DatasetThumbnail from './DatasetThumbnail.vue'
import { capitalize } from 'lodash-es'
import {
  datasetVisibilityQuery,
  deleteDatasetQuery,
  reprocessDatasetQuery,
} from '../../../api/dataset'
import { mdTypeSupportsOpticalImages } from '../../../lib/util'
import { encodeParams } from '../../Filters/index'
import reportError from '../../../lib/reportError'
import safeJsonParse from '../../../lib/safeJsonParse'
import { plural } from '../../../lib/vueFilters'
import ElapsedTime from '../../../components/ElapsedTime'
import DownloadDialog from './DownloadDialog'

function removeUnderscores(str) {
  return str.replace(/_/g, ' ')
}

export default {
  name: 'DatasetItem',
  components: {
    DatasetInfo,
    DatasetThumbnail,
    ElapsedTime,
    DownloadDialog,
  },
  filters: {
    plural,
  },
  props: ['dataset', 'currentUser', 'idx', 'hideGroupMenu'],
  data() {
    return {
      showMetadataDialog: false,
      disabled: false,
      deferRender: this.idx >= 20,
      showDownloadDialog: false,
    }
  },

  computed: {
    opticalImageAlignmentHref() {
      return {
        name: 'add-optical-image',
        params: { dataset_id: this.dataset.id },
      }
    },

    isOpticalImageSupported() {
      return mdTypeSupportsOpticalImages(this.$store.getters.filter.metadataType)
    },

    formatSubmitter() {
      const { name } = this.dataset.submitter
      return name
    },

    formatDatasetName() {
      return this.dataset.name
    },

    analyzerType() {
      return this.dataset.analyzer.type
    },

    getDate() {
      const unknown = ''
      if (!this.dataset.id) {
        return unknown
      }

      const fields = this.dataset.id.split('_')
      if (fields.length < 2) {
        return unknown
      }

      const date = fields[0]
      const time = fields[1].split('m')[0].replace('h', ':')
      return `${date} ${time}`
    },

    metadata() {
      const datasetMetadataExternals = {
        Submitter: this.dataset.submitter,
        PI: this.dataset.principalInvestigator,
        Group: this.dataset.groupApproved ? this.dataset.group : null,
        Projects: this.dataset.projects,
      }
      return Object.assign(safeJsonParse(this.dataset.metadataJson), datasetMetadataExternals)
    },

    metaboliteDatabases() {
      const dbs = this.dataset.molDBs
      if (typeof dbs === 'string') {
        return [dbs]
      } else {
        return dbs
      }
    },

    formatOrganism() {
      return removeUnderscores(this.dataset.organism)
    },

    formatCondition() {
      return removeUnderscores(this.dataset.condition).toLowerCase()
    },

    formatOrganismPart() {
      return removeUnderscores(this.dataset.organismPart).toLowerCase()
    },

    formatResolvingPower() {
      const rpSection = this.metadata.MS_Analysis.Detector_Resolving_Power
      const { mz, Resolving_Power: rp } = rpSection
      return (rp / 1000).toFixed(0) + 'k @ ' + mz
    },

    canEdit() {
      if (this.currentUser != null) {
        if (this.currentUser.role === 'admin') {
          return true
        }
        if (
          this.currentUser.id === this.dataset.submitter.id
          && !['QUEUED', 'ANNOTATING'].includes(this.dataset.status)
        ) {
          return true
        }
      }
      return false
    },

    canDelete() {
      return this.canEdit && !this.isPublished
    },

    canEditOpticalImage() {
      return this.canEdit && this.dataset.status === 'FINISHED'
    },

    canReprocess() {
      return this.currentUser != null
         && this.currentUser.role === 'admin'
    },

    editHref() {
      return {
        name: 'edit-metadata',
        params: { dataset_id: this.dataset.id },
      }
    },

    disabledClass() {
      return this.disabled ? 'ds-item-disabled' : ''
    },

    visibilityText() {
      if (this.datasetVisibility != null) {
        const { submitter, group, projects } = this.datasetVisibility
        const submitterName = this.currentUser && submitter.id === this.currentUser.id ? 'you' : submitter.name
        const all = [
          submitterName,
          ...(group ? [group.name] : []),
          ...(projects || []).map(p => p.name),
        ]
        return 'These annotation results are not publicly visible. '
          + `They are visible to ${all.join(', ')} and METASPACE Administrators.`
      }
      return null
    },

    isPublished() {
      return this.dataset.projects.filter(p => p.publicationStatus === 'PUBLISHED').length > 0
    },
  },
  async created() {
    // Defer rendering of most elements until after the first render, so that the page becomes interactive sooner
    const delayFrames = Math.floor(this.idx / 10)
    try {
      for (let i = 0; i < delayFrames; i++) {
        await new Promise(resolve => requestAnimationFrame(resolve))
      }
    } catch (err) { /* Browser/test doesn't support requestAnimationFrame? */ }
    this.deferRender = false
  },
  apollo: {
    datasetVisibility: {
      query: datasetVisibilityQuery,
      skip: true,
      variables() {
        return { id: this.dataset.id }
      },
    },
  },

  methods: {
    resultsHref(databaseName) {
      const filter = Object.assign({}, this.$store.getters.filter, {
        database: databaseName,
        datasetIds: [this.dataset.id],
      })
      return {
        path: '/annotations',
        query: Object.assign({},
          encodeParams(filter, '/annotations', this.$store.state.filterLists),
          { mdtype: this.dataset.metadataType }),
      }
    },

    showMetadata(e) {
      e.preventDefault()
      this.showMetadataDialog = true
    },

    addFilter(field) {
      const filter = Object.assign({}, this.$store.getters.filter)
      if (field === 'polarity') {
        filter.polarity = capitalize(this.dataset.polarity)
      } else if (field === 'submitter') {
        filter[field] = this.dataset.submitter.id
      } else if (field === 'group') {
        filter[field] = this.dataset.group.id
      } else {
        filter[field] = this.dataset[field] || this[field]
      }
      this.$store.commit('updateFilter', filter)
      this.$emit('filterUpdate', filter)
    },

    async openDeleteDialog() {
      const force = this.currentUser != null
         && this.currentUser.role === 'admin'
         && this.dataset.status !== 'FINISHED'
      try {
        const msg = `Are you sure you want to ${force ? 'FORCE-DELETE' : 'delete'} ${this.formatDatasetName}?`
        await this.$confirm(msg, {
          type: force ? 'warning' : null,
          lockScroll: false,
        })
      } catch (cancel) {
        return
      }

      try {
        this.disabled = true
        const resp = await this.$apollo.mutate({
          mutation: deleteDatasetQuery,
          variables: {
            id: this.dataset.id,
            force,
          },
        })
        this.$emit('datasetMutated')
      } catch (err) {
        this.disabled = false
        reportError(err, 'Deletion failed :( Please contact us at contact@metaspace2020.eu')
      }
    },

    async handleReprocess() {
      try {
        this.disabled = true
        await this.$apollo.mutate({
          mutation: reprocessDatasetQuery,
          variables: {
            id: this.dataset.id,
            force: true,
          },
        })
        this.$notify.success('Dataset sent for reprocessing')
        this.$emit('datasetMutated')
      } catch (err) {
        reportError(err)
      } finally {
        this.disabled = false
      }
    },

    handleDropdownCommand(command) {
      if (command.startsWith('filter_')) {
        this.addFilter(command.substring('filter_'.length))
      } else if (command === 'view_group') {
        this.$router.push({
          name: 'group',
          params: {
            groupIdOrSlug: this.dataset.group.id,
          },
        })
      }
    },

    formatFdrLevel() {
      return this.dataset.fdrCounts.levels.join(', ')
    },

    formatFdrCounts() {
      return this.dataset.fdrCounts.counts.join(', ')
    },

    formatDbName() {
      return this.dataset.fdrCounts.dbName
    },

    loadVisibility() {
      this.$apollo.queries.datasetVisibility.start()
    },
  },
}
</script>

<style lang="scss">
  .opt-image--container {
    padding: 10px 0 10px 10px;
    margin: 0px;
    flex: none;
  }

 .dataset-item {
   border-radius: 5px;
   // Can't use box-sizing:border-box due to IE11 flexbox limitations, so instead using `calc(100% - 2px)`
   flex: 1 1 calc(100% - 2px);
   min-height: 120px;
   min-width: 600px;
   max-width: 950px;
   margin: 3px;
   padding: 0px;
   display: flex;
   flex-direction: row;
   justify-content: space-between;
   transition: 0.2s cubic-bezier(.4, 0, .2, 1);
   transition-property: box-shadow;
   font-variant-numeric: proportional-nums;
 }

 .dataset-item:hover {
   box-shadow: 0 2px 12px 0 rgba(0, 0, 0, 0.1);
 }

 .ds-info {
   padding: 10px;
   margin: 0px;
   flex-grow: 1;
   min-width: 0%; /* This may seem pointless, but it's necessary to prevent text overflowing: https://css-tricks.com/flexbox-truncated-text/ */
 }

 .ds-actions {
   padding: 10px 22px 10px 0px;
   margin: 0px;
   flex: none;
   display: flex;
   flex-direction: column;
 }

 .ds-add-filter {
   cursor: pointer;
   font-weight: 500;
 }

 .striped-progressbar {
   height: 14px;
   border-radius: 2px;
   margin: 3px 0;
   width: 100%;
   background-size: 30px 30px;
   background-image: linear-gradient(135deg,
     rgba(255, 255, 255, .30) 25%, transparent 25%, transparent 50%,
     rgba(255, 255, 255, .30) 50%,
     rgba(255, 255, 255, .30) 75%, transparent 75%, transparent);

   animation: animate-stripes 3s linear infinite;
 }

 @keyframes animate-stripes {
   0% {background-position: 0 0;} 100% {background-position: 60px 0;}
 }

 .processing {
   background-color: lightgreen;
 }

 .queued {
   background-color: lightblue;
 }

 .db-link-list {
   font-size: initial;
 }

 .ds-item-line {
   overflow: hidden;
   white-space: nowrap;
   text-overflow: ellipsis;
 }

</style>
