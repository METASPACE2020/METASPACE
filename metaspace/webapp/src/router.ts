/* eslint-disable vue/max-len */
import Vue, { AsyncComponent } from 'vue'
import VueRouter, { RawLocation } from 'vue-router'
import AboutPage from './modules/App/AboutPage.vue'
import DatasetsPage from './modules/Datasets/DatasetsPage.vue'
import { DialogPage, ResetPasswordPage } from './modules/Account'
import { redirectAfterSignIn } from './modules/Account/signInReturnUrl'
import { updateDBParam } from './modules/Filters/url'
import NotFoundPage from './modules/App/NotFoundPage.vue'

Vue.use(VueRouter)

const asyncPagesFreelyTyped = {
  AnnotationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "AnnotationsPage" */ './modules/Annotations/AnnotationsPage.vue'),
  DatasetSummary: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetSummary" */ './modules/Datasets/summary/DatasetSummary.vue'),
  MetadataEditPage: () => import(/* webpackPrefetch: true, webpackChunkName: "MetadataEditPage" */ './modules/MetadataEditor/MetadataEditPage.vue'),
  ImageAlignmentPage: () => import(/* webpackPrefetch: true, webpackChunkName: "ImageAlignmentPage" */ './modules/ImageAlignment/ImageAlignmentPage.vue'),
  UploadPage: () => import(/* webpackPrefetch: true, webpackChunkName: "UploadPage" */ './modules/MetadataEditor/UploadPage.vue'),
  DesignPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DesignPage" */ './design/DesignPage.vue'),

  // These pages are relatively small as they don't have any big 3rd party dependencies, so pack them together
  DatasetTable: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Datasets/list/DatasetTable.vue'),
  HelpPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/HelpPage.vue'),
  EditUserPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/UserProfile/EditUserPage.vue'),
  CreateGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/GroupProfile/CreateGroupPage.vue'),
  ProjectsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Project/ProjectsListPage.vue'),
  SystemHealthPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Admin/SystemHealthPage.vue'),
  GroupsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Admin/GroupsListPage.vue'),
  PrivacyPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/PrivacyPage.vue'),
  TermsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/TermsPage.vue'),
  PublicationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/PublicationsPage.vue'),

  // These pages use sanitizeHtml, which is big
  ViewGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ './modules/GroupProfile/ViewGroupPage.vue'),
  ViewProjectPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ './modules/Project/ViewProjectPage.vue'),
}
const asyncPages = asyncPagesFreelyTyped as Record<keyof typeof asyncPagesFreelyTyped, AsyncComponent>

const convertLegacyUrls = () => {
  const { pathname, hash, search } = window.location
  if (pathname === '/' && hash && hash.startsWith('#/')) {
    history.replaceState(undefined, undefined as any, hash.slice(1))
  }
  if (pathname === '/annotations') {
    const updatedQueryString = updateDBParam(search.slice(1))
    if (updatedQueryString !== null) {
      history.replaceState(undefined, undefined as any, `${pathname}?${updatedQueryString}`)
    }
  }
}
convertLegacyUrls()

const router = new VueRouter({
  mode: 'history',
  routes: [
    { path: '/(about)?', component: AboutPage, meta: { footer: true, headerClass: 'bg-primary' } },
    { path: '/annotations', component: asyncPages.AnnotationsPage },
    {
      path: '/datasets',
      component: DatasetsPage,
      children: [
        { path: '', component: asyncPages.DatasetTable },
        { path: 'summary', component: asyncPages.DatasetSummary },
      ],
    },
    { path: '/datasets/edit/:dataset_id', name: 'edit-metadata', component: asyncPages.MetadataEditPage },
    { path: '/datasets/:dataset_id/add-optical-image', name: 'add-optical-image', component: asyncPages.ImageAlignmentPage },
    { path: '/upload', component: asyncPages.UploadPage },
    { path: '/help', component: asyncPages.HelpPage, meta: { footer: true } },
    { path: '/user/me', component: asyncPages.EditUserPage },

    { path: '/admin/health', component: asyncPages.SystemHealthPage },
    { path: '/admin/groups', component: asyncPages.GroupsListPage },

    { path: '/account/sign-in', component: DialogPage, props: { dialog: 'signIn' } },
    { path: '/account/sign-in-success', redirect: redirectAfterSignIn },
    { path: '/account/create-account', component: DialogPage, props: { dialog: 'createAccount' } },
    { path: '/account/forgot-password', component: DialogPage, props: { dialog: 'forgotPassword' } },
    { path: '/account/reset-password', component: ResetPasswordPage },

    { path: '/group/create', component: asyncPages.CreateGroupPage },
    { path: '/group/:groupIdOrSlug', name: 'group', component: asyncPages.ViewGroupPage },

    { path: '/project/:projectIdOrSlug', name: 'project', component: asyncPages.ViewProjectPage },
    { // Legacy URL sent in "request access" emails up until Feb 2019
      path: '/project/:projectIdOrSlug/manage',
      redirect: { path: '/project/:projectIdOrSlug', query: { tab: 'members' } } as RawLocation,
    },
    { path: '/projects', component: asyncPages.ProjectsListPage },

    { path: '/terms', component: asyncPages.TermsPage, meta: { footer: true } },
    { path: '/privacy', component: asyncPages.PrivacyPage, meta: { footer: true } },
    { path: '/publications', component: asyncPages.PublicationsPage, meta: { footer: true } },

    { path: '/design', component: asyncPages.DesignPage, meta: { footer: true, flex: true } },

    { path: '*', component: NotFoundPage, meta: { footer: true, flex: true } },
  ],
  scrollBehavior(to, from, savedPosition) {
    if (to.hash) {
      return { selector: to.hash, offset: { x: 0, y: 64 } } // offset header height
    }
    if (savedPosition) {
      return savedPosition
    }
    if (to.path !== from.path) {
      return { x: 0, y: 0 }
    }
  },
})

const { href } = router.resolve({ name: 'project', params: { projectIdOrSlug: 'REMOVE' } }, undefined, true)
export const PROJECT_URL_PREFIX = location.origin + href.replace('REMOVE', '')

export default router
