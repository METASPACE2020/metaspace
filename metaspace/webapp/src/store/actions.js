import { fetchOptionListsQuery } from '../api/metadata'
import { decodeParams } from '../modules/Filters'
import { computed } from 'vue'
import store from '../store/index'

export default {
  async initFilterLists(context, apolloClient) {
    if (context.state.filterListsLoading || context.state.filterLists != null) return
    await context.commit('setFilterListsLoading')

    const response = await apolloClient.query({
      query: fetchOptionListsQuery,
      fetchPolicy: 'cache-first',
    })

    // set annotationIds default according to values passed from snapshot to the store state
    context.commit('setFilterLists', {
      ...response.data,
      annotationIds: computed(() => store.state.snapshotAnnotationIds),
    })

    // Refresh the current filter so that computed defaults that depend on `filterLists` are applied
    const filter = decodeParams(context.state.route, context.state.filterLists)

    context.commit('updateFilter', filter)
  },
}
