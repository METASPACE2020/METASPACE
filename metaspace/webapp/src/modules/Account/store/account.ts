import { Module } from 'vuex';
import { DialogType } from '../dialogs';
import router from '../../../router';

export interface AccountState {
  dialog: DialogType | null;
  showDialogAsPage: boolean;
}

const account = {
  namespaced: true,
  state: {
    dialog: null,
    showDialogAsPage: false
  },
  mutations: {
    showDialog(state: AccountState, dialog: DialogType) {
      state.dialog = dialog;
      state.showDialogAsPage = false;
    },
    showDialogAsPage(state: AccountState, dialog: DialogType) {
      state.dialog = dialog;
      state.showDialogAsPage = true;
    },
    hideDialog(state: AccountState, dialog: DialogType) {
      if (state.dialog === dialog) {
        state.dialog = null;
        if (state.showDialogAsPage) {
          router.push('/');
        }
      }
    }
  }
};

export default account as Module<AccountState, any>;
