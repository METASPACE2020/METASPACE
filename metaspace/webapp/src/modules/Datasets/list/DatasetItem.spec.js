import { mount } from '@vue/test-utils'
import Vue from 'vue'
import Vuex from 'vuex'
import router from '../../../router'
import store from '../../../store'
import { sync } from 'vuex-router-sync'
import DatasetItem from './DatasetItem.vue'

Vue.use(Vuex)
sync(store, router)

describe('DatasetItem', () => {
  const user = { id: 'user' }
  const submitter = { id: 'submitter' }
  const admin = { id: 'admin', role: 'admin' }
  const dataset = {
    id: 'mockdataset',
    status: 'FINISHED',
    metadataJson: JSON.stringify({
      Metadata_Type: 'Imaging MS',
      MS_Analysis: {
        Detector_Resolving_Power: { mz: 1234, Resolving_Power: 123456 },
      },
    }),
    databases: [
      { name: 'CHEBI', id: 2 },
      { name: 'HMDB-v2.5', id: 6 },
      { name: 'HMDB-v4', id: 22 },
    ],
    polarity: 'POSITIVE',
    fdrCounts: {
      databaseId: 6,
      dbName: 'HMDB-v2.5',
      levels: [10],
      counts: [20],
    },
    groupApproved: true,
    submitter,
    organism: 'organism',
    organismPart: 'organismPart',
    condition: 'condition',
    analyzer: {
      type: 'type'
    },
    projects: [],
    uploadDT: '2020-04-17T11:37:50.318667'
  }
  const unpublished = { name: 'project', publicationStatus: 'UNPUBLISHED' }
  const underReview = { name: 'project', publicationStatus: 'UNDER_REVIEW' }
  const published = { name: 'project', publicationStatus: 'PUBLISHED' }

  it('should be able to delete if unpublished', () => {
    const propsData = {
      currentUser: submitter,
      dataset: {
        ...dataset,
        projects: [unpublished]
      },
    }
    const wrapper = mount(DatasetItem, { router, store, propsData })
    expect(wrapper).toMatchSnapshot()
  })

  it('should not show the publication status if cannot edit', () => {
    const propsData = {
      currentUser: user,
      dataset: {
        ...dataset,
        projects: [published]
      },
    }
    const wrapper = mount(DatasetItem, { router, store, propsData })
    expect(wrapper).toMatchSnapshot()
  })

  it('should not show the publication status if processing', () => {
    const propsData = {
      currentUser: submitter,
      dataset: {
        ...dataset,
        projects: [published],
        status: 'ANNOTATING'
      }
    }
    const wrapper = mount(DatasetItem, { router, store, propsData })
    expect(wrapper).toMatchSnapshot()
  })

  it('should show "Under review" status', () => {
    const propsData = {
      currentUser: submitter,
      dataset: {
        ...dataset,
        projects: [underReview]
      }
    }
    const wrapper = mount(DatasetItem, { router, store, propsData })
    expect(wrapper).toMatchSnapshot()
  })

  it('should show "Published" status', () => {
    const propsData = {
      currentUser: submitter,
      dataset: {
        ...dataset,
        projects: [published]
      }
    }
    const wrapper = mount(DatasetItem, { router, store, propsData })
    expect(wrapper).toMatchSnapshot()
  })

  it('should prefer "Published" status', () => {
    const propsData = {
      currentUser: submitter,
      dataset: {
        ...dataset,
        projects: [published, underReview]
      }
    }
    const wrapper = mount(DatasetItem, { router, store, propsData })
    expect(wrapper).toMatchSnapshot()
  })

  it('should show admin options', () => {
    const propsData = {
      currentUser: admin,
      dataset: {
        ...dataset,
        projects: [published]
      }
    }
    const wrapper = mount(DatasetItem, { router, store, propsData })
    expect(wrapper).toMatchSnapshot()
  })
})
