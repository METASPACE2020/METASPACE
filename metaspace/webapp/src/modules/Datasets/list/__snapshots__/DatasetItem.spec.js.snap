// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasetItem should match snapshot 1`] = `
<div class="dataset-item border border-solid border-gray-200 leading-5">
  <mock-el-dialog title="Provided metadata">
    <div class="el-row">
      <mock-el-tree id="metadata-tree" data="[object Object],[object Object]" node-key="id" default-expanded-keys="MS analysis,Detector resolving power,Data Management"></mock-el-tree>
    </div>
  </mock-el-dialog>
  <div class="opt-image--container"><a href="/datasets/mockdataset/add-optical-image" class="">
      <div title="Add Optical Image" class="thumb thumb__editable thumb__empty">
        <!---->
        <!----> <img src="../../../assets/no_opt_image.png" alt="Add optical image" class="thumb--img">
        <div class="thumb--overlay el-icon-edit el-icon-plus"></div>
      </div>
    </a></div>
  <div class="ds-info">
    <div class="ds-item-line flex"><span class="font-bold truncate"></span>
      <mock-el-popover trigger="hover" placement="top" class="ml-1">
        <div slot-key="default">
          <div mock-v-loading="true">

          </div>
        </div>
        <div slot-key="reference"><i class="el-icon-lock"></i></div>
      </mock-el-popover>
    </div>
    <div class="ds-item-line text-gray-700"><span title="Filter by species" class="ds-add-filter">
        organism</span>,
      <span title="Filter by organism part" class="ds-add-filter">
        organismpart</span> <span title="Filter by condition" class="ds-add-filter">
        (condition)</span></div>
    <div class="ds-item-line"><span title="Filter by ionisation source" class="ds-add-filter">
        </span> +
      <span title="Filter by analyzer type" class="ds-add-filter">
        type</span>,
      <span title="Filter by polarity" class="ds-add-filter">
        positive mode</span>,
      RP 123k @ 1234
    </div>
    <div class="ds-item-line">
      Submitted <mock-elapsed-time date="2020-04-17T11:37:50.318667"></mock-elapsed-time> by
      <span title="Filter by submitter" class="ds-add-filter">
        </span>
      <!---->
    </div>
    <div class="ds-item-line"><span><a href="/annotations?db=HMDB-v2.5&amp;ds=mockdataset" class="">20 annotations</a>
        @ FDR 10% (HMDB-v2.5)
      </span></div>
  </div>
  <div class="ds-actions"><span><i class="el-icon-picture"></i> <mock-el-popover trigger="hover" placement="top"><div slot-key="default"><div class="db-link-list">
          Select a database:
          <div><a href="/annotations?db=HMDB-v2.5&amp;ds=mockdataset" class="">
              HMDB-v2.5
            </a></div><div><a href="/annotations?db=HMDB-v4&amp;ds=mockdataset" class="">
              HMDB-v4
            </a></div><div><a href="/annotations?db=CHEBI&amp;ds=mockdataset" class="">
              CHEBI
            </a></div></div> </div><div slot-key="reference"><a>Browse annotations</a></div></mock-el-popover> <br></span>
    <!---->
    <!---->
    <div><i class="el-icon-view"></i> <a href="#">Show full metadata</a></div>
    <div><i class="el-icon-edit"></i> <a href="/datasets/edit/mockdataset" class="">
        Edit metadata
      </a></div>
    <!---->
    <div class="ds-delete"><i class="el-icon-delete"></i> <a href="#" class="text-danger">Delete dataset</a></div>
    <!---->
  </div>
  <!---->
</div>
`;
